src/puppetlabs/trapperkeeper/bootstrap.clj

```diff
@@ -132,26 +136,102 @@
             (str "Unable to find bootstrap.cfg file via --bootstrap-config "
                  "command line argument, current working directory, or on classpath")))))
 
-(schema/defn chain-files :- [schema/Str]
-  "Takes a list of files, reads all their lines in, and returns a flattened seq
-   of all their lines put together.
+(schema/defn indexed
+  "Returns seq of [index, item] pairs
+  [:a :b :c] -> ([0 :a] [1 :b] [2 :c])"
+  [coll]
+  (map vector (iterate inc 0) coll))
 
-   In the case of an empty file, nil would be returned by line-seq, so we remove
-   all the nils at the end"
-  [files :- [(schema/protocol io/IOFactory)]]
-  (remove nil? (flatten (map #(line-seq (io/reader %)) files))))
+(schema/defn get-annotated-bootstrap-entries :- [AnnotatedBootstrapEntry]
+  [configs :- [(schema/protocol io/IOFactory)]]
+  (flatten (for [config configs]
+             (for [[line-number line-text] (indexed (map remove-comments (line-seq (io/reader config))))
+                   :when (not (empty? line-text))]
+               {:bootstrap-file (str config)
+                :line-number (inc line-number)
+                :entry line-text}))))
+
+(defn find-duplicates
+  "Collects duplicates base on running f on each item.
+   Returns a map where the keys will be the result of running f on each item,
+   and the values will be lists of items that are duplicates of eachother"
+  [coll f]
+  (->> coll
+       (group-by f)
+       ; filter by map value with only 1 item
+       (remove #(= 1 (count (val %))))))
+
+(schema/defn duplicate-protocol-error
+  "Returns an IllegalArgumentException with a nice error message
+   saying in what file and line each service entry was found"
+  [protocol-id :- schema/Keyword
+   duplicate-services :- [(schema/protocol services/ServiceDefinition)]
+   service->entry-map :- {(schema/protocol services/ServiceDefinition) AnnotatedBootstrapEntry}]
+  (let [make-error-message (fn [service]
+                             (let [entry (get service->entry-map service)]
+                               (format "%s:%s\n%s"
+                                       (:bootstrap-file entry)
+                                       (:line-number entry)
+                                       (:entry entry))))]
+    (IllegalArgumentException.
+      (format (str "Duplicate implementations found for service protocol '%s':\n%s")
+              protocol-id
+              (string/join "\n" (map make-error-message duplicate-services))))))
+
+(schema/defn check-duplicate-service-implementations
```

Comment:
probably name this with a trailing `!` to indicate that it has side effects (throws)